File: wrappers/claude/interactive/delivery_test.go
// SPDX-License-Identifier: MIT

package interactive

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"io"
	"net"
	"testing"

	kit "github.com/antst/sessionbus/bus/sdk/go"
)

type writeConn struct {
	net.Conn
	write func([]byte) (int, error)
}

func (c writeConn) Write(b []byte) (int, error) { return c.write(b) }

func TestNativeWriteCapturesExactFrameAndCloses(t *testing.T) {
	a, b := net.Pipe()
	defer b.Close()
	frames := make(chan map[string]any, 1)
	closed := make(chan struct{})
	go func() {
		r := bufio.NewReader(b)
		line, _ := r.ReadBytes('\n')
		var f map[string]any
		_ = json.Unmarshal(line, &f)
		frames <- f
		_, _ = r.ReadByte()
		close(closed)
	}()
	receipt, err := DeliverNative(Recipient{SessionID: "captured-id", Socket: "native", Context: context.Background()}, kit.DeliveryRequest{MessageID: "message", From: kit.DeliverySource{SessionID: "sender"}, Body: "exact body"}, func(context.Context, string, string) (net.Conn, error) { return a, nil })
	if err != nil || receipt.Disposition != "written" {
		t.Fatalf("%#v %v", receipt, err)
	}
	f := <-frames
	if f["session_id"] != "captured-id" || f["from"] != "sender" || f["msg_id"] != "message" || f["msgV"] != float64(1) || f["message"].(map[string]any)["content"] != "exact body" {
		t.Fatal(f)
	}
	<-closed
}
func TestNativePreSubmissionAndUncertainBoundaries(t *testing.T) {
	for _, kind := range []string{"missing", "cancelled", "connect", "partial", "write-error", "cancel-during-write"} {
		t.Run(kind, func(t *testing.T) {
			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()
			captured := Recipient{SessionID: "id", Socket: "native", Context: ctx}
			if kind == "missing" {
				captured.Socket = ""
			}
			if kind == "cancelled" {
				cancel()
			}
			writes := 0
			dials := 0
			dial := func(context.Context, string, string) (net.Conn, error) {
				dials++
				if kind == "connect" {
					return nil, errors.New("connect failure")
				}
				a, b := net.Pipe()
				t.Cleanup(func() { _ = a.Close(); _ = b.Close() })
				return writeConn{Conn: a, write: func(data []byte) (int, error) {
					writes++
					switch kind {
					case "partial":
						return len(data) - 1, nil
					case "cancel-during-write":
						cancel()
						return 0, io.ErrClosedPipe
					default:
						return 0, errors.New("write failure")
					}
				}}, nil
			}
			r, err := DeliverNative(captured, kit.DeliveryRequest{}, dial)
			if kind == "missing" || kind == "cancelled" || kind == "connect" {
				if err != nil || r.Disposition != "rejected" || writes != 0 {
					t.Fatalf("%#v %v writes%d", r, err, writes)
				}
				if kind != "connect" && dials != 0 {
					t.Fatal("unavailable delivery dialed")
				}
			} else {
				var p *kit.ProtocolError
				if !errors.As(err, &p) || p.Code != -32603 || string(p.Data) != `"uncertain_submission"` || r.Disposition != "" || writes != 1 {
					t.Fatalf("%#v %v writes%d", r, err, writes)
				}
			}
		})
	}
}

File: docs/designs/claude-0.5.0/node-reference/delivery.test.mjs
// SPDX-License-Identifier: MIT
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { EventEmitter, once } from 'node:events';
import { createServer } from 'node:net';
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { ProtocolError, validate } from '@sessionbus/kit';
import { deliverNative } from './delivery.mjs';

const request={message_id:'original',from:{session_id:'sender@host',product:'test',groups:['g']},body:'unaltered\nbody'};
function fixture() {
  const socket=new EventEmitter(); let callback, bytes, destroyed=false;
  socket.write=(b,cb)=>{bytes=b;callback=cb;}; socket.destroy=()=>{destroyed=true;};
  const controller=new AbortController(); const context={session_id:'old-id',socket:'/test',signal:controller.signal};
  return {socket,controller,context,bytes:()=>bytes,callback:e=>callback(e),destroyed:()=>destroyed};
}
test('one actual local socket write returns written without waiting for native EOF',async t=>{
  const dir=mkdtempSync(join(tmpdir(),'claude-delivery-')),path=join(dir,'s');
  let receive; const received=new Promise(r=>receive=r);
  const server=createServer(s=>s.once('data',b=>{receive(JSON.parse(b));s.destroy();}));
  server.listen(path); await once(server,'listening');
  t.after(async()=>{await new Promise(r=>server.close(r));rmSync(dir,{recursive:true,force:true});});
  const result=await deliverNative({session_id:'actual',socket:path,signal:new AbortController().signal},request);
  assert.deepEqual(result,{disposition:'written'}); const frame=await received;
  assert.deepEqual(frame,{msgV:1,msg_id:'original',type:'user',priority:'next',from:'sender@host',session_id:'actual',message:{role:'user',content:request.body}});
});
test('captured target/source remain fixed across delayed write completion and rename',async()=>{
  const f=fixture(); const result=deliverNative(f.context,{...request,from:{...request.from,name:'Real sender'}},()=>f.socket);
  f.context.session_id='new-id'; f.socket.emit('connect');
  assert.equal(JSON.parse(f.bytes()).session_id,'old-id'); assert.equal(JSON.parse(f.bytes()).from,'Real sender');
  f.callback(); assert.deepEqual(await result,{disposition:'written'}); assert.ok(f.destroyed());
});
test('missing socket, connect error and cancellation before submission are truthful rejections',async()=>{
  for(const kind of ['missing','connect','cancel']) {
    const f=fixture(); if(kind==='missing') f.context.socket='';
    const result=deliverNative(f.context,request,()=>f.socket);
    if(kind==='connect') f.socket.emit('error',new Error('controlled')); if(kind==='cancel') f.controller.abort();
    assert.equal((await result).disposition,'rejected'); assert.equal(f.bytes(),undefined);
  }
});
test('partial write, callback error, clear/supersession cancellation and EOF are uncertain after submission',async()=>{
  for(const event of ['callback','abort','end','close','error']) {
    const f=fixture(), result=deliverNative(f.context,request,()=>f.socket); f.socket.emit('connect');
    if(event==='callback') f.callback(new Error('partial')); else if(event==='abort') f.controller.abort(); else f.socket.emit(event,new Error('transport'));
    await assert.rejects(result,e=>e instanceof ProtocolError && e.code===-32603 && e.data==='uncertain_submission' && validate('RPCError',{code:e.code,message:e.message,data:e.data}));
    assert.ok(f.destroyed());
  }
});
test('complete write is not reclassified by a subsequent identity cancellation or EOF',async()=>{
  const f=fixture(), result=deliverNative(f.context,request,()=>f.socket); f.socket.emit('connect'); f.callback(); f.controller.abort(); f.socket.emit('end');
  assert.deepEqual(await result,{disposition:'written'});
  assert.ok(validate('MessageSendResult',{message_id:'m',deliveries:[{target:'target',session_id:'target',delivery_id:'d',disposition:'written'}]}));
  assert.ok(validate('MessageSendResult',{message_id:'m',deliveries:[{target:'target',session_id:'target',delivery_id:'d',disposition:'rejected',reason:'no_receipt'}]}));
});
